docs/source/blueprints/BP004.rst

```diff
@@ -0,0 +1,211 @@
+
+A dependency graph is a directed acyclic graph describing the dependencies between the nodes. In a dependency graph, each arc of a node connects to the node on which it depends. The diagram below illustrates a dependency graph:
+
+.. image:: images/BP004/dependency_graph.png
+   :width: 400
+   :alt: A dependency graph
+
+In the example above, node A depends on nodes B and C; node B depends on nodes D and E; node C depends on nodes E and F. Nodes D, E and F are the independent nodes and take the initial positions. A dependency graph may have multiple topological orders. Some valid topological orders for the graph above are:
+
+1. A, C, B, F, E, D;
+
+2. A, C, F, B, E, D;
+
+3. A, B, D, C, F, E.
+
+The inverse topological order represents the execution order for the tasks, where independent nodes can execute in parallel. The algorithm to create the topological order removes the source nodes (those nodes with in-degree equal to zero) while enumerating them. With the removal of a source node, new source nodes may be created. The algorithm ends when there are no more nodes to remove, meaning the graph is empty or the remaining nodes are not sources. In the latter case, the graph has cycles and cannot be handled as a dependency graph.
+
+A dependency graph allows the construction and control of more elaborated structures representing the dependencies between the tests, requirements, and others. One example covered by the dependency graph is a test depending on a plugin, like SysInfo, which can run before and after a test or before and after a job.
+
+What is not covered by the Dependency Graphs
+============================================
+
+The main objective of the dependency graph is to solve the problem where two tasks depending on the same requirement, in the same environment, create two different requirements tasks that may conflict.
+
+What this blueprint does not solve is how the dependencies, after fulfilled, are cached in each kind of environment. For example, when two different tests have the same dependency and run inside a container, it is the caching system responsibility to cache that state of the container and pass it to the tests before they run.
+
+Specification
+*************
+
+On the default runner, the requirements resolution happens during the creation of the test task. The ``nrunner runner`` plugin creates a requirement task when a test has a requirement listed on its structure and sets the new test task to depend on this requirement task. As mentioned, different tests with the same requirement will depend on different requirements tasks that perform the same activity. See `runner_nrunner.py#L207 <https://github.com/avocado-framework/avocado/blob/09221f33ac2ff974a076f93601e8bd7fd6738e7a/avocado/plugins/runner_nrunner.py#L207>`__ for more details on how it happens.
```

Comment:
nrunner runner ? Maybe nrunner_runner ?

Reply:
In the first sentence, I'm naming the plugin, not the file; that is why I added a space between nrunner and runner. I understand that it is called `nrunner_runner` because we do not want to add spaces on a file name.

Reply:
Based on our previous documents, and the plugin name itself I find confusing. Specially because is using the "code" syntax.

Reply:
Could you point me to a document where it calls a plugin like `some_plugin`? I could not find any occurrence on the documentation. I wanted to highlight the term, but I can remove it and let it be in plain text if that makes the text less confusing, no problem.

Reply:
Btw, not a blocker, just a comment. I could not find reference to "nrunner runner" plugin there:

 `grep plugin docs/* -R | grep \` | grep runner`.
 
 Looks like we do have different forms there. But my point is that the plugin is called "runner_nrunner.py" I could not find "nrunner_runner.py".